### DISCUSSION TITLE
[API Request] Discussion and clarification for how track-only artists should be handled in legacy endpoints

### DISCUSSION THREAD

**Discussion by dweymouth** [opensubsonic/open-subsonic-api#61] [API requests]
### Type of change

API Clarification

### Proposal description

We currently have a divergence between Gonic and LMS on how track-only artists are handled. Gonic does not assign IDs to track-only artists or return them in the `getArtists` response. This is also the behavior of Navidrome and Airsonic-Advanced (which have not implemented OpenSubsonic). This presents the problem that track-only artists are not navigable, but this is how original Subsonic handled it. LMS on the other hand assigns IDs to all artists and includes all the artists's album appearances as part of their discography. This has the problem of not distinguishing an artists's albums (that they are credited as an album artist on) from the albums that they just appear on.

This can all be handled with additional APIs or API extensions, but that is a separate discussion. This is just to discuss what the behavior with the existing APIs should be - I'd suggest going with Gonic's approach as it's closest to original Subsonic and is thererfore not a breaking change for existing clients.

### Backward compatibility

- [ ] No backward compatibility impact.

### API details

Only album artists by default should be included in the `getArtists` response, and only the albums that the artist has an Album Artist credit for should be included in the albums list from the `getArtist` API. We can discuss further extensions to these APIs in other discussions, eg including a new section in the `getArtist` response to list the appearances, and new arguments to `getArtists` to include track-only artists or not.

**Comment by Tolriq** [comment]
>  Gonic does not assign IDs to track-only

**It does now** that's the whole point of the issue ....

Since it returns "artists" [] to songs they must have ids if they have ids they exist, if they exist they must be returned in getArtist / Search and be accessible via getArtistInfo.

If he only want to returns albumArtists then he only return that field https://opensubsonic.netlify.app/docs/responses/child/

But now he is in violation of OpenSubsonic.

I really don't understand why and how you both conclude that this is a breaking change and that hiding data that would break other servers and clients is not breaking.
When a server does only support and handle albumArtist then it returns what it have, if a server support track Artists then it returns what it have the API was designed like that and OpenSubsonic definition with the roles support this.

There's not only track artists, there's Composers, musicians, and all the possible roles.

So again IMO, not returning all artists in the endpoints for artists is the breaking change.

**If an artist exist and have an ID it must be accessible via the endpoints. I don't even understand why there's a discussion here.**

If an artist does not have an ID and is only used as display value then it's something else, but unlike your first statement this is no more the case now.

**And this is what happens in Navidrome too, any artist that is returned with an ArtistID is accessible via the artist endpoints and the search3 and get the getArtistInfo as it should.**

**Reply by epoupon** [reply]
I do agree with @Tolriq  on this. All artists should be returned in search and getArtists endpoints. We have far more gain with this on all existing clients: using the legacy API you can already list your favorite composer and see on which albums they appear.
If this really bothers you to have too many artists, you can make a setting server side to tweak this (so that users have the choice), and we can extend the OS API to filter by role

**Reply by sentriz** [reply]
> If an artist exist and have an ID it must be accessible via the endpoints. I don't even understand why there's a discussion here.
> And this is what happens in Navidrome too, any artist that is returned with an ArtistID is accessible via the artist endpoints and the search3 and get the getArtistInfo as it should.

yes! i agree it must be accessible.. but it's a big deal for a server to suddenly start returning 1000s of extra track artists when a user wants to view their artists, or search for their favourite artist. it's a really big change and hurts user experience for some. 

also, by opensubsonic redefining getArtists or search3 to operate on album artists, it makes it really hard for servers to be opensubsonic compatible: how can navidrome, gonic, airsonic-advanced ever implement track artists and be opensubsonic compaltible, without changing their existing user experience? they can't.

can we just define role filter? where clients can request all types of artists. if no filter is provided, LSM could default to all types, gonic and existing servers could default to albumartist, then all existing user experience remains the same

navidrome even mention this in their docs: 
![image](https://github.com/opensubsonic/open-subsonic-api/assets/6832539/bf791a4e-4929-4fcc-9365-cc6e080a6518)

it's a big deal to some users to change this (if/when navidrome/airsonic implement track artists)

**Reply by Tolriq** [reply]
> also, by opensubsonic redefining getArtists or search3 to operate on album artists, it makes it really hard for servers to be opensubsonic compatible: how can navidrome, gonic, airsonic-advanced ever implement track artists and be opensubsonic compaltible, without changing their existing user experience? they can't.

So to resume you want to add support for track artist without adding support for track artist, the current API is not defined as albumArtist even in Subsonic, it's just that most do not support track artist at all.

> yes! i agree it must be accessible.

So don't expose them with your current implementation as you are breaking things for current clients. If the current API does not fit your need, refine the API before breaking it by forcing your opposite point of view that breaks clients and the purpose of this project.
Returning "artists" array for songs that are track artists with ID and those artists not being accessible currently breaks things and user experience. Since no one can use those values without being OpenSubsonic compatible don't expose them until you don't break things.

>  it's a really big change and hurts user experience for some.

Removing that for those who actually want it and currently use it is also a big changes and hurt experience ...


As we say in French you want "le beurre, l’argent du beurre, et le cul de la crémière"

**Reply by sentriz** [reply]
you're whole argument of "the only reason servers don't return track artists over getArtists is because they don't support them yet" is misplaced. it's a very intentional decision  for servers to return album artists there because otherwise getArtists get cluttered. it's a selling point of servers even (again, see navidrome and gonic readmes). lots of users do not want to see track artists where by default. i have not had one gonic issue open where a user requested to see track artists in their artist list

again, i still agree it's useful to make them more accessable somehow. so let's add a role filter? why not? you haven't really replied to this yet

and for gonic, yes track artists are useful. you can click on a track artist and see their bio with getArtistInfo. if they have any full albums where they're credited as an album artist you'll see those too. maybe in future we can extened this to return their singles too

**Reply by Tolriq** [reply]
> . so let's add a role filter? why not? you haven't really replied to this yet

I've answered already like 4 times ... So one more time, because this is a quick unthought hack to a more global need ... There's a lot more needs around artists filtering and searching and composers, and the many other roles ... This is not just a track artist issue here.
So adding a quick filter that will block us in a couple of days when we decide to define the proper API for the proper needs is definitively not the purpose of OpenSubsonic, already one breaking change on your side, let's avoid generating more and make this project worse than the initial situation?

> lots of users do not want to see track artists where by default.

Then add a setting in your server like others instead of breaking the API to fit your need ? ;)

Anyway this is going nowhere, you have made your decision. You wanted to rush this and forced pushed before finding a solution, your choice, the result is that why should I fight for OpenSubsonic to exist when it's failed at it first addition ....

**Reply by sentriz** [reply]
> So one more time, because this is a quick unthought hack to a more global need ... There's a lot more needs around artists filtering and searching and composers, and the many other roles ... This is not just a track artist issue here.

i'm not suggesting to do any quick hacks. let's do it properly. there's no rush. why not in this discussion here? lets open a conversation

> breaking the API to fit your need ? ;)

my whole point is that opensubsonic is where the breaking change is. in redefining getArtists to be all artists, where most existing implementations chose explicitly to be album artist. i keep saying this but i'm not really being understood

**Reply by Tolriq** [reply]
> in redefining getArtists to be all artists, where most existing implementations chose explicitly to be album artist

Again no, most/all implementations do not return artistId for those artists so there's no issue here .. Does Airsonic return artistId for those songs since this is the Bible for you? Or does it return the artistID of the album?
One server does it as defined.

Now you want to return IDs unlike the others but also want to change the API do your definition of it. You are mixing things here.
Most servers do not return artistIDs for track IDs to have the API not return those track artist, not the other way around like you want to imply.

**Reply by sentriz** [reply]
it's fine to return an artist ID that doesn't appear (by default) in getArtists or search3. they are still navigable and useful with getArtist/getArtistInfo

the only place it becomes an issue is the offline sync case

but we can improve the situation as i've said before: we can define some nice role filters and extensions together, and/or some dump.view for the offline case

**Reply by Tolriq** [reply]
> it's fine to return an artist ID that doesn't appear (by default) in getArtists or search3. 

Your decision, what server does that since your logic is always that it's how other servers work?

Why your new decision would be the API definition, while what is currently in prod by other servers would finally become bad?

> the only place it becomes an issue is the offline sync case

No it's not ... It's an issue for current servers like LMS that do return all artists .... You change the way things works for your need and make current servers behavior invalid by doing so ...

**Reply by sentriz** [reply]
again, in my opionion, it's not that i'm breaking opensubsonic, or changing how things work (i am ofc in some ways) but more importantly, i'm saying that opensubsonic is wrong in the first place to define getArtists to mean all artists. since that is not the status quo

**Reply by Tolriq** [reply]
So one server does it correctly, all the API naming says something, no server do return artistID for track ID.

But somehow, in your opinion your new need generates  a status quo that became what you decide it should be even if nowhere anything match your choice ?

So again sorry to disappoint you but if there's a status quo it's what I try to explain since the start, you don't agree and want your way on your server it's totally legit, but don't try to make it appear like it's how Subsonic works and OpenSubsonic is breaking things here .

**Reply by sentriz** [reply]
were going around in circles. maybe let me ask an honest question:

what do you think will happen if tomorrow navidrome or airsonic (lots and lots of users) decide they want to support track artists and opensubsonic. how will this affect their getArtists and search3 views? and how will the 1000s of users react to the change?

**Reply by Tolriq** [reply]
They will do it when the API have added the filter to onlyShowAlbumArtist? Or they will offer a toggle in their server setting?

Why are you trying now to invoke future potential things? Do you think you can justify breaking things by making supposition that other servers will break things too and not find a proper solution and accept the API for that it is?

So the honest question is: Why do you absolutely want to justify your changes by saying the others breaks the API I do it right, instead of just saying it's my server I do what I want ....

This discussion will always be present whatever API improvement we make it would be based on the current status quo not what you want it to be, so if the improvement is adding a onlyAlbumArtist filter, you will be against because it's a breaking change for you ...

So no we are not going in circles we are going in opposite directions, that's what I try to explain, by diverging you create future breaking change that won't be possible to fix without breaking the others. Others that are already in prod ....

**Reply by sentriz** [reply]
> Why are you trying now to invoke future potential things? Do you think you can justify breaking things by making supposition that other servers will break things too and not find a proper solution and accept the API for that it is?

> So the honest question is: Why do you absolutely want to justify your changes by saying the others breaks the API I do it right, instead of just saying it's my server I do what I want ....

because it's important that servers have an easy path forward to be opensubsonic compatible, and changing getArtists/search3 is a huge breaking change for existing servers. we want all servers to opensubsonic compatible right? whats the point otherwise

> They will do it when the API have added the filter to onlyShowAlbumArtist? Or they will offer a toggle in their server setting?

again, above, to be compatible, have an easy path forward for big and small servers, and not upset many many users, i really think all artists should be opt in, not out. if the setting is opt out, then all existing clients immediately break by not knowing about onlyShowAlbumArtist or whatever the filter may be. the most popular client like dsub is rarely updated for example. this is important

**Reply by Tolriq** [reply]
That's the thing they don't break..... They return more data, they now support more artists they return more artists this is logical and expected even if you don't want to see the other side ...

But right now you break LMS and the future evolutions...

Why break everything and not have a server setting for the users who don't want the data on legacy clients.
You add support for track artist. It's there. The user don't want to have them have an option to remove them if you want to please them.

Don't break the API and the future for your very own opinion ...

Servers do transcoding per client, they can offer legacy default setting per client too ....
Servers have settings to fit user needs, that's the point.

**Reply by sentriz** [reply]
> That's the thing they don't break..... They return more data, they now support more artists they return more artists this is logical and expected even if you don't want to see the other side ...
> Why break everything and not have a server setting for the users who don't want the data on legacy clients.
You add support for track artist. It's there. The user don't want to have them have an option to remove them if you want to please them.

they do break. again it's a selling point that servers handle album artists correctly and don't clutter artists lists. do you think other servers will suddently start returning 1000s of more artists that the user hasn't heard of?

if the user explicitly wants them that's fine. and i think it should be controlled over the api on a per request setting. not a global server side setting

**Reply by Tolriq** [reply]
If your server have a selling point to hide stuff then hide it by default with a setting. Stop justifying your choice by what others may do in the future and ignoring what users do today ..... 

So fun to talk to a wall. I give up, you're thinking right I'm wrong and do not understand anything. Good luck it's not worth it.

**Comment by dweymouth** [comment]
> > Gonic does not assign IDs to track-only
> 
> **It does now** that's the whole point of the issue ....

**Ooohhh**.. my apologies, I didn't realize that this was the change that Gonic made. I thought he was including track-only artists in the list but with a name only and no ID - which IMO is fine and good since clients could then treat them as non-navigable and display a plain label instead of a hyperlink. In other words they only exist as a display value in tracklist views. This is how Navidrome does it which of course hasn't implemented OpenSubsonic yet, but a (single-valued) track artist that has no albums will have its name set in the `artist` property of the track's `Child` object but the `artistID` will be unset.

I see why Tolriq is upset @sentriz  - I agree now that IF the artist has an ID (ie it *exists* as a meaningful entity in the DB) then it is something he will want to sync into the offline DB and it should be reteieved by the endpoints including `getArtists` and `search3`. If you want to exclude track artists I think not assigning them IDs at all is preferable, until later when more API extensions enable the proper handling of track artists as a distinct thing from album artists.

Now for @epoupon's decision - I think it's also not optimal that the discography lists are cluttered with "appearance" albums for artists. If it were up to me, we would say that the album list returned in `getArtist` is only for albums that they're credited as an album artist on, and we'll have to wait for API extensions to properly expose the "appearances".

**Reply by sentriz** [reply]
> which IMO is fine and good since clients could then treat them as non-navigable and display a plain label instead of a hyperlink. In other words they only exist as a display value in tracklist views. This is how Navidrome does it which of course hasn't implemented OpenSubsonic ye

small side note: they are still navigable with getArtist and getArtistInfo (just not getArtists and search3) 

> not optimal that the discography lists are cluttered with "appearance" albums for artists

agree :+1: :+1:

**Reply by dweymouth** [reply]
> small side note: they are still navigable with getArtist and getArtistInfo (just not getArtists and search3)

Yeah I think that is Tolriq's issue. If they're not navigable at all no problem, but if they *are*, then it's something he needs to sync with Symfonium's offline DB so if it's not retrievable in any endpoint that iterates through everything, then that's a problem

**Reply by sentriz** [reply]
i get that, that why i suggested some extra param to getArtists that retrieves all sorts of artists. that way we can keep the behaviour of getArtists not being cluttered with appearances. or maybe a dump.view that returns everything instead of the `search3?query=*` hack

but i dunno seems like we're not getting anywhere with this line of thinking

**Reply by dweymouth** [reply]
Is there any chance you would hold off on track artists until we have the extension to `getArtists` to filter by roles? That seems like it would make everyone happy. Or I wonder how Tolriq would feel about track artists with no IDs in the OpenSubsonic response (I for one am totally OK with that since I already handle that case for Navidrome)

**Reply by sentriz** [reply]
i'd be happy to implement a role filter for get artists so long as the default without a filter is album artists. because think of old and very popular clients like dsub. they will likely never implement a role filter. and it we start returning all appearances then they get flooded with appearances 

and i think its not a big deal to have some logic for offline clients (who are not aware of the old filter yet) like "can't find track artist in artists response -> don't make it navigable"

that way legacy clients can still navigate appearance artists who also happen to have full albums. this happens a lot in my library

**Reply by dweymouth** [reply]
> that way legacy clients can still navigate appearance artists who also happen to have full albums. this happens a lot in my library

Navidrome handles this too (of course with single-valued tags only) but they assign an ID to track artists that do have albums, and only omit the ID for track-**only** artists

**Reply by dweymouth** [reply]
Also, separately to the offline sync issue, it seems weird to me to make an artist page with no content navigable. (ie a track-only artist, it will have no content unless the getTopSongs API is returning songs for that artist)

**Reply by sentriz** [reply]
> Navidrome handles this too (of course with single-valued tags only) but they assign an ID to track artists that do have albums, and only omit the ID for track-only artists

ah yes that's a good point and a good compromise. also means that getArtists is the whole set of album artists with nothing missing

ill consider this :+1: 


---

but in future i think it would be nice opensubsonic extension for clients to explictly request/search all types of artists, navigate to them, then see their individual tracks on compilation albums. like a new `songs` field in the artist response

**Reply by Tolriq** [reply]
> but i dunno seems like we're not getting anywhere with this line of thinking

The I don't give a shit about the current API usage and servers, I have a new feature I want it my way so I want all the others to comply or break ? ;) Yes that usually don't get anywhere ...

Again: 100% of the servers that return artistsID allow finding those artist in getArtists and Search3. You want to go your way, assume but this is the reality of the current situation. Now how you want to present it.

>  Or I wonder how Tolriq would feel about track artists with no IDs in the OpenSubsonic response

I'm just a simple client participant now, but the API state that the IDs are mandatory in the openSubsonic part, so I would advice against as it would break all the clients that relies on the schema and expect the field to be present.

**Reply by sentriz** [reply]
> The I don't give a shit about the current API usage and servers

no, it's the "the i don't want my users to be flooded with 1000s of appearances in the artist lists" as i've said already

**Reply by dweymouth** [reply]
Alright last try at being a mediator here... @Tolriq, given that sentriz will probably not implement exactly what you want right now, would you prefer it a) as it is now, or b) with IDs omitted from track-only artists so that they exist in display name only and not as a DB entity? Keeping in mind this can **all be resolved** happily once we have additional APIs or extensions to properly deal with roles filtering

**Reply by Tolriq** [reply]
"I" "MY USERS" do you see a pattern here? ... It's your users, your server add a server option. Don't impose your needs to the others and breaking what is currently already in prod.

But as I said I no more care, I support 10 different media providers and already have code path per Subsonic servers, anything you do have no real impact on me, just on the global OpenSubsonic concept and future. This is time lost to try to discuss.

**Reply by sentriz** [reply]
it's not just me. it's gonic, navidrome, airsonic who all currently operate on album artists. and it will be an issue if they want to start supporting track artists with legacy clients who can't filter. their lists will be flooded

(this may be the case for ampache, funkwhale, and others - haven't checked)

but yes this is going nowhere so ill probably add a sever side option

**Comment by certuna** [comment]
I've encountered the same issue when adding track artist support to Navidrome (see [this PR](https://github.com/navidrome/navidrome/pull/2583))
IMO there's two ways to do it:
- API allows to query albumCount, i.e. the client can ask for all artists where albumCount > 0
- no changes to the API, Client receives all artists, and before displaying them to users, drops those with albumCount = 0

The only real difference is bandwidth efficiency - you send over tons of artists that the client immediately discards.

**Reply by Tolriq** [reply]
Unfortunately not all servers handle the albumCount the same for track artists.
And some clients do not discard and do use the data.

The question here is about the definition of the API and handling of legacy clients. OpenSubsonic already expose the necessary data to filter without albumCount with the artist roles.

From the API naming and how it's done and some server currently do in prod, the API define that getArtists returns all artists even if some want to read it the way that better suit their own need.

If some servers support track artists and composers and use legacy clients they can offer a setting to by default only return album artists until the OS filtering solution is found like LMS do for example.

The discussion for the new filters is done at https://github.com/opensubsonic/open-subsonic-api/discussions/57

So yes a new filter is needed for OS to better handle the roles, but the question here is about how to handle this currently without touching the API. Adding random quick hacks will just block the proper filtering.

**Reply by certuna** [reply]
Yes I know - the current Subsonic API, clients just ask for all artists and get all artists, it's up to the server to define what it considers "all artists". I think that's fine (it's how I've implemented it in Navidrome, if @deluan agrees with the approach: a serverside `AlbumArtistsOnly = true/false` setting). In the end it's up to the servers to set albumCount correctly (i.e. zero for artists without albums), and up to clients to filter them out if they want.

**Comment by Fingel** [comment]
It's super difficult to read through all the flames here to see if there was ever a resolution to this. My use case is simple: I want to display a list of _album_ Artists, without it being cluttered by individual artists from collaborations, etc. Is this possible with the OpenSubsonic API? The closest I got was to iterate over the "artists" list on AlbumID3WithSongs but even then I get duplicate artists (weirdly, the same artist name with multiple IDs). 
Anyone found a way?
https://github.com/Fingel/gelly/issues/100

**Reply by Tolriq** [reply]
Check the roles and filter client side.

**Reply by Fingel** [reply]
Thanks for the response. I'm still pretty confused. I found the [ArtistID3](https://opensubsonic.netlify.app/docs/responses/artistid3/) response which I think you are referring to, which has the roles list. But I can't find a list endpoint that actually uses it. [getArtists](https://opensubsonic.netlify.app/docs/endpoints/getartists/) does not have the presence of roles. [getArtist](https://opensubsonic.netlify.app/docs/endpoints/getartist/) does, but that would mean requiring a request for each artist - not acceptable. 

Even assuming there is a list endpoint that includes roles, I'm trying to figure out how I'd use that to get the album artist for an album. I'd have to store the collection of artists in memory, maybe filtered down to those that have the "albumArtist" role, check the `artistId` field of every album in [getAlbumlist2](https://opensubsonic.netlify.app/docs/endpoints/getalbumlist2/) and see if it is present in the artist collection?

I also noticed the [getArtists](https://opensubsonic.netlify.app/docs/endpoints/getartists/) api does include an album count field (but no roles). I think I read somewhere else that people use this to filter artists with < 1 album out? I'll give that a try.
